Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean

```diff
@@ -757,6 +771,124 @@ def toConstProdLinearMap : (V1 →ᵃ[k] V2) ≃ₗ[R] V2 × (V1 →ₗ[k] V2) w
 
 end Module
 
+section Pi
+
+-- The pi construction for AffineMaps. This accepts dependent types in both the module V
+-- and the AffineSpace P. This leads to pretty unreadable / un-loogle-able type signatures,
+-- so a non-dependent pi construction pi' is also given with corresponding theorems.
+
+variable {ι : Type*} {φv φp : ι → Type*} [(i : ι) → AddCommGroup (φv i)]
+  [(i : ι) → Module k (φv i)] [(i : ι) → AffineSpace (φv i) (φp i)]
+
+/-- The dependently-typed version of `AffineMap.pi`, where the module `V1 := φv i` and the
+ AffineSpace `P2 := φp i` can depend on the index `i : ι`.
+-/
+def pi (f : (i : ι) → (P1 →ᵃ[k] φp i)) : P1 →ᵃ[k] ((i : ι) → φp i) where
+  toFun := fun m a ↦ (f a).1 m
+  linear := LinearMap.pi (fun a ↦ (f a).2)
+  map_vadd' := fun _ _ ↦ funext fun a ↦ (f a).3 _ _
+
+--fp for when the image is a dependent AffineSpace φp i, fv for when the
+--image is a Module φv i, f' for when the image isn't dependent.
+variable (fp : (i : ι) → (P1 →ᵃ[k] φp i)) (fv : (i : ι) → (P1 →ᵃ[k] φv i))
+  (f' : ι → P1 →ᵃ[k] P2)
+
+@[simp]
+theorem pi_apply (c : P1) (i : ι) : pi fp c i = fp i c :=
+  rfl
+
+theorem pi_comp {V3 P3 : Type*} [AddCommGroup V3] [Module k V3] [AffineSpace V3 P3]
+    (g : P3 →ᵃ[k] P1) : (pi fp).comp g = pi (fun i => (fp i).comp g) :=
+  rfl
+
+theorem pi_eq_zero : pi fv = 0 ↔ ∀ i, fv i = 0 := by
+  simp only [AffineMap.ext_iff, Function.funext_iff, pi_apply]
+  exact forall_comm
+
+theorem pi_zero : pi (fun _ ↦ 0 : (i : ι) → P1 →ᵃ[k] φv i) = 0 := by
+  ext; rfl
+
+theorem proj_pi (i : ι) : (proj i).comp (pi fp) = fp i :=
+  ext fun _ => rfl
+
+/-- An ι-indexed collection of AffineMaps `P1 →ᵃ P2` lifts to an AffineMap from P1 to the
+  Module of functions, `ι → P2`. This module is `Pi.module`. This is the AffineMap analog
+  of `LinearMap.pi`. Actually defeq to `AffineMap.pi` but the type signature is
+  much easier to read.
+-/
+def pi' : (ι → (P1 →ᵃ[k] P2)) → P1 →ᵃ[k] (ι → P2) := pi
+
+@[simp]
+theorem pi'_apply (c : P1) (i : ι) : pi' f' c i = f' i c :=
+  rfl
+
+theorem pi'_comp {V3 P3 : Type*} [AddCommGroup V3] [Module k V3] [AffineSpace V3 P3]
+    (g : P3 →ᵃ[k] P1) : (pi' f').comp g = pi' (fun i => (f' i).comp g) :=
+  rfl
+
+theorem pi'_eq_zero (f : (i : ι) → P1 →ᵃ[k] V2) : pi' f = 0 ↔ ∀ i, f i = 0 :=
+  pi_eq_zero f
+
+theorem pi'_zero : pi' (fun _ ↦ 0 : (i : ι) → P1 →ᵃ[k] V2) = 0 := by ext; rfl
+
+theorem proj_pi' (i : ι) : (proj i).comp (pi' f') = f' i :=
+  ext fun _ => rfl
+
+/-- The `AffineMap` version of `LinearMap.single`, `AddMonoidHom.single` and `Pi.single`. -/
+def single [DecidableEq ι] (i : ι) : φv i →ᵃ[k] (j : ι) → φv j :=
+  { toFun := Pi.single i
+    linear := LinearMap.single i
+    map_vadd' := fun _ _ ↦ Pi.single_add i _ _
+  }
```

Comment:
I don't think this is worth having; we should already have that every linear map is also an affine map

Reply:
dropped